server/src/main/java/org/elasticsearch/rest/action/search/HttpChannelTaskHandler.java

```diff
@@ -0,0 +1,158 @@
+            //otherwise, given that its listener has already been called, the task id would never be removed.
+            if (link.getAndIncrement() == 0) {
+                TaskId taskId = new TaskId(client.getLocalNodeId(), task.getId());
+                closeListener.registerTask(httpChannel, taskId);
+            }
+        }
+
+        //TODO test case where listener is registered, but no tasks have been added yet:
+        // - connection gets closed, channel will be removed, no tasks will be cancelled
+
+        //TODO check that no tasks are left behind through assertions at node close
+    }
+
+    final class CloseListener implements ActionListener<Void> {
+        private final Client client;
+        final Set<TaskId> taskIds = new CopyOnWriteArraySet<>();
+        private final AtomicReference<HttpChannel> channel = new AtomicReference<>();
+
+        CloseListener(Client client) {
+            this.client = client;
+        }
+
+        void registerTask(HttpChannel httpChannel, TaskId taskId) {
+            if (channel.compareAndSet(null, httpChannel)) {
+                //In case the channel is already closed when we register the listener, the listener will be immediately executed which will
+                //remove the channel from the map straight-away. That is why we do this in two stages. If we provided the channel at close
+                //listener initialization we would have to deal with close listeners calls before the channel is in the map.
+                httpChannel.addCloseListener(this);
+            }
+            this.taskIds.add(taskId);
+        }
+
+        private void unregisterTask(TaskId taskId) {
+            this.taskIds.remove(taskId);
+        }
+
+        @Override
+        public void onResponse(Void aVoid) {
+            //When the channel gets closed it won't be reused: we can remove it from the map as there is no chance we will
+            //register another close listener to it later.
+            //The channel reference may be null, if the connection gets closed before we set it.
+            //The channel must be found in the map though as this listener gets registered after the channel is added.
+            //TODO test channel null here? it can happen!
+            httpChannels.remove(channel.get());
+            for (TaskId previousTaskId : taskIds) {
+                CancelTasksRequest cancelTasksRequest = new CancelTasksRequest();
+                cancelTasksRequest.setTaskId(previousTaskId);
+                //We don't wait for cancel tasks to come back. Task cancellation is just best effort.
+                //Note that cancel tasks fails if the user sending the search request does not have the permissions to call it.
+                client.admin().cluster().cancelTasks(cancelTasksRequest, ActionListener.wrap(r -> {}, e -> {}));
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            //nothing to do here
```

Comment:
I don't know what situation can yields this but we should also remove the channel from the map here to be on the safe side (no leak).

Reply:
oh well, I just realized that I am never removing the channel from the map, it's an oversight as I haven't written a test for when a connection gets closed yet, will fix.